### PR TITLE
feat(crypto): CRP-2420 add new signature verification package initially supporting canister signatures

### PR DESCRIPTION
**Pull request by przydatek**
Setup `ic-signature-verification` package, which is intended to be a standalone crate with minimal dependencies
for verifying signatures supported by the Internet Computer.

Add the first utility function to the package, namely `verify_canister_sig()` for verifying [canister signatures](https://internetcomputer.org/docs/current/references/ic-interface-spec#canister-signatures). 
More comprehensive tests are needed, and they will be added in a follow-up PR.

### CODE REVIEW

**Comment by przydatek** [comment]
Hi @fspreiss , 
In addition to the more comprehensive tests, there could be (at least) two more extensions/changes (maybe as follow-up PRs, or future work)

1. Argument types: right now we have mostly &[u8], but maybe we could wrap these e.g. as `CanisterSigPublicKeyDer([u8])`,  or `CanisterSignatureCbor([u8])`, so that it is not so easy to put the arguments in a wrong order.
2. The return value is just `Result<(), String>`, which on one hand is nice because it is simple/generic, but maybe a more elaborate error-enum could be introduced, maybe not for the main function, but for internal reporting between the helpers, and for a potential additional API to be consumed internally by the replica code.

Wdyt?

**Review by marko-k0 — APPROVED** [review]

File: packages/ic-signature-verification/src/lib.rs
Comment (fspreiss) [inline packages/ic-signature-verification/src/lib.rs:20]: General note: later we'll want to use this function also directly in the replica code. There, however, the DER is already stripped away from the public key. I guess we don't have to solve this right now, but eventually we need a way to deal with this.

Another relevant aspect for using this code in the replica is _caching_: currently the replica does caching on the verified BLS signatures. Eventually we want to have that here as well, at least optionally. Most likely, the way to achieve this is to add caching capabilities to `ic_verify_bls_signature`.

**Review by fspreiss — COMMENTED** [review]
Thanks a lot, @przydatek: I didn't check all verification details yet, but this PR looks really good! ~Above~ Below you find my comments so far. The code is very concise (partially enabled through the use of `ic-certification`) and nice to read.

> Hi @fspreiss , In addition to the more comprehensive tests, there could be (at least) two more extensions/changes (maybe as follow-up PRs, or future work)
> 
> 1. Argument types: right now we have mostly &[u8], but maybe we could wrap these e.g. as `CanisterSigPublicKeyDer([u8])`,  or `CanisterSignatureCbor([u8])`, so that it is not so easy to put the arguments in a wrong order.

You are right in that having multiple/only `&[u8]` as parameter types is suboptimal. Introducing newtype wrappers like `CanisterSigPublicKeyDer` would help, but ideally these types are somewhat smart because being able to instantiate a struct called `CanisterSigPublicKeyDer` with invalid DER is also suboptimal. The good thing in our situation is that if a caller passes the arguments in the wrong order, nothing bad can happen: the verification will simply fail, which might mean that the `&[u8]` parameters are good enough. Still, at least for the use of the library in the replica code, dedicated types might help. We can also discuss offline how to proceed here.

> 2. The return value is just `Result<(), String>`, which on one hand is nice because it is simple/generic, but maybe a more elaborate error-enum could be introduced, maybe not for the main function, but for internal reporting between the helpers, and for a potential additional API to be consumed internally by the replica code.

Agreed. `String` is nice and simple, but indeed we want to use this library later in the CryptoComponent/Replica code (instead of the current canister sig verification code) and for that a proper error type may actually be good to have, at least internally. Anyway, this is also [what Roman would recommend](https://mmapped.blog/posts/12-rust-error-handling#prefer-specific-enums).

File: packages/ic-signature-verification/Cargo.toml
Comment (przydatek) [reply packages/ic-signature-verification/Cargo.toml:1]: Done.

File: packages/ic-ledger-hash-of/CHANGELOG.md
Comment (przydatek) [reply packages/ic-ledger-hash-of/CHANGELOG.md:13]: oops, no idea how that happened, sorry!  Fixed now.

**Review by przydatek — COMMENTED** [review]
Thanks, PTAL.

File: packages/ic-signature-verification/Cargo.toml
Comment (gregorydemay) [inline packages/ic-signature-verification/Cargo.toml:2]: Apologies if this was discussed earlier, but the difference with the existing [ic-crypto-standalone-sig-verifier](https://github.com/dfinity/ic/blob/4c75016a882cd113dd41e924382b469deff7093c/rs/crypto/standalone-sig-verifier/Cargo.toml#L2) is not immediately clear to me, since this was also intended as a standalone sig verifier. Would it make sense to merge both?

Comment (przydatek) [reply packages/ic-signature-verification/Cargo.toml:2]: This new package (unlike the existing one) does not depend  on ic-repo-types, and can be published on crates.io

**Review by fspreiss — COMMENTED** [review]
I now double-check the verification logic agains the [spec](https://internetcomputer.org/docs/current/references/ic-interface-spec#canister-signatures), and this looks very good.

File: packages/ic-signature-verification/Cargo.toml
Comment (gregorydemay) [reply packages/ic-signature-verification/Cargo.toml:2]: I see, thanks for the explanation!

File: packages/ic-signature-verification/src/lib.rs
Comment (fspreiss) [reply packages/ic-signature-verification/src/lib.rs]: I mainly find it odd that one library (`ic-certification`) implements the types, and another library (`ic-cbor`) implements the parsing. My impression is it would be best to have this all handled transparently directly in `ic-certification` (which may use `ic-cbor` underneath if need be). Maybe @nathanosdev has some insights here.

I'm OK with keeping things as if for now.

Comment (nathanosdev) [reply packages/ic-signature-verification/src/lib.rs]: The main purpose for `ic-cbor` was to provide a deserialization that would keep a smaller bundle size at the end when compiling to WASM and binding to JavaScript. `serde_cbor` added quite a significant increase to file size.

This was important for the Service Worker, which we wanted to keep as small as possible, but since the Service Worker is gone now, we don't have that requirement anymore. I'd be in favor of not using it and eventually removing it in favor of `serde_cbor`.

File: packages/ic-signature-verification/src/lib.rs
Comment (przydatek) [reply packages/ic-signature-verification/src/lib.rs]: The reason for not parsing the certificate "outside" of the helper is that the certified_data-check refers to the parts of the same `signature`-object, so it feels "safer" to do all the checks inside of the helper.  Agreed though that sig-path-check is a different check, so I did split the helper.
Also, `verify_certificate` is more costly then the checks, so postponing it to after the checks can potentially save some cycles.
Please take a look at the slightly changed flow -- if you prefer, I can still change to the structure you've suggested.

**Review by przydatek — COMMENTED** [review]
Thanks, PTAL.

File: packages/ic-signature-verification/src/lib.rs
Comment (fspreiss) [reply packages/ic-signature-verification/src/lib.rs]: We can go with what we have now. If `ic_canister_sig_creation` is extended later, we can then also make the necessary adaptions here. But until then, this should not affect/block merging of this PR here.

**Review by fspreiss — APPROVED** [review]
Thanks for the various changes, @przydatek!

File: packages/ic-signature-verification/src/lib.rs
Comment (przydatek) [reply packages/ic-signature-verification/src/lib.rs]: Done.

**Review by przydatek — COMMENTED** [review]
Thanks, will merge once green.

**Pull request merged**
